### PR TITLE
Use blocked dofmap and allow setting of PETSc Mat type

### PR DESCRIPTION
**Pull request by garth-wells**
This PR exploits elements with a block size to support block CSR matrices and to reduce the memory storage for dofmaps and related objects.

Block size isn't yet exploited in DirichletBC as it a bit more complicated.

### CODE REVIEW

File: cpp/dolfinx/function/Function.h [hunk 1]
```diff
@@ -268,28 +268,37 @@ class Function
     std::shared_ptr<const fem::FiniteElement> element
         = _function_space->element();
     assert(element);
-    const int block_size = element->block_size();
+    // const int block_size = element->block_size();
```
Comment (michalhabera) [inline cpp/dolfinx/function/Function.h:271]: Leftovers

File: cpp/dolfinx/function/Function.h [hunk 2]
```diff
@@ -268,28 +268,37 @@ class Function
...
+    const int space_dimension = element->space_dimension() / bs_element;
 
     // If the space has sub elements, concatenate the evaluations on the sub
     // elements
     const int num_sub_elements = element->num_sub_elements();
-    if (num_sub_elements > 1 && num_sub_elements != block_size)
+    if (num_sub_elements > 1 and num_sub_elements != bs_element)
     {
-      if (block_size != 1)
+      std::cout << "Oooops" << std::endl;
```
Comment (michalhabera) [inline cpp/dolfinx/function/Function.h:287]: Leftover

File: cpp/dolfinx/function/FunctionSpace.cpp [hunk 3]
```diff
@@ -44,12 +44,9 @@ internal_tabulate_dof_coordinates(
   std::shared_ptr<const common::IndexMap> index_map = dofmap->index_map;
   assert(index_map);
 
-  int bs = dofmap->index_map_bs();
-  int element_block_size = element->block_size();
-
-  std::int32_t local_size
-      = bs * (index_map->size_local() + index_map->num_ghosts())
-        / element_block_size;
+  // int bs = dofmap->index_map_bs();
```
Comment (michalhabera) [inline cpp/dolfinx/function/FunctionSpace.cpp:47]: Leftover

File: cpp/dolfinx/la/SparsityPattern.cpp [hunk 4]
```diff
@@ -34,7 +34,8 @@ SparsityPattern::SparsityPattern(
     const std::vector<std::vector<const SparsityPattern*>>& patterns,
     const std::array<std::vector<std::pair<
                          std::reference_wrapper<const common::IndexMap>, int>>,
-                     2>& maps)
+                     2>& maps,
+    const std::array<std::vector<int>, 2>& bs)
```
Comment (michalhabera) [inline cpp/dolfinx/la/SparsityPattern.cpp:38]: It takes block sizes but sets `_bs({1, 1})` below?

File: python/CMakeLists.txt [hunk 5]
```diff
@@ -4,6 +4,8 @@ PROJECT(dolfinx_pybind11)
 
 find_package(DOLFINX REQUIRED)
 
+set(CMAKE_BUILD_TYPE DEBUG)
```
Comment (michalhabera) [inline python/CMakeLists.txt:7]: Unrelated change

File: python/demo/elasticity/demo_elasticity.py [hunk 6]
```diff
@@ -188,7 +163,7 @@ def sigma(v):
 solver.setOperators(A)
 
 # Compute solution
-solver.setMonitor(lambda ksp, its, rnorm: print("Iteration: {}, rel. residual: {}".format(its, rnorm)))
+# solver.setMonitor(lambda ksp, its, rnorm: print("Iteration: {}, rel. residual: {}".format(its, rnorm)))
```
Comment (michalhabera) [inline python/demo/elasticity/demo_elasticity.py:166]: Leftover

File: python/demo/stokes-taylor-hood/demo_stokes-taylor-hood.py [hunk 7]
```diff
@@ -323,11 +322,10 @@ def lid_velocity_expression(x):
 ksp_p.getPC().setType("jacobi")
 
 # Monitor the convergence of the KSP
-opts = PETSc.Options()
+# opts = PETSc.Options()
 # opts["ksp_monitor"] = None
 # opts["ksp_view"] = None
-
-ksp.setFromOptions()
+# ksp.setFromOptions()
```
Comment (michalhabera) [inline python/demo/stokes-taylor-hood/demo_stokes-taylor-hood.py:328]: Leftovers

File: python/dolfinx/fem/assemble.py [hunk 8]
```diff
@@ -306,7 +309,11 @@ def _(A: PETSc.Mat,
         for j, a_sub in enumerate(a_row):
             if a_sub is not None:
                 Asub = A.getLocalSubMatrix(is_rows[i], is_cols[j])
-                assemble_matrix(Asub, a_sub, bcs, diagonal)
+                # Asub.setBlockSizes(a_sub.function_spaces[0].dofmap.bs,
```
Comment (michalhabera) [inline python/dolfinx/fem/assemble.py:312]: Leftovers

File: python/dolfinx/wrappers/nls.cpp [hunk 9]
```diff
@@ -13,12 +13,12 @@
 #include <pybind11/pybind11.h>
 #include <pybind11/stl.h>
 
-#ifdef DEBUG
+// #ifdef DEBUG
```
Comment (michalhabera) [inline python/dolfinx/wrappers/nls.cpp:16]: Why this?

File: python/test/unit/fem/test_assembler.py [hunk 10]
```diff
@@ -209,7 +209,11 @@ def test_assemble_manifold():
     assert numpy.isclose(A.norm(), 25.0199)
 
 
-@pytest.mark.parametrize("mode", [dolfinx.cpp.mesh.GhostMode.none, dolfinx.cpp.mesh.GhostMode.shared_facet])
+@pytest.mark.parametrize("mode",
+                         [
+                             dolfinx.cpp.mesh.GhostMode.none,
```
Comment (michalhabera) [inline python/test/unit/fem/test_assembler.py:214]: Non-standard formatting

File: cpp/dolfinx/fem/DofMapBuilder.cpp [hunk 11]
```diff
@@ -550,20 +548,15 @@ DofMapBuilder::build(MPI_Comm comm, const mesh::Topology& topology,
...
+      dofmap[local_dim0 * cell + j] = new_node;
     }
   }
   assert(dofmap.rows() % node_graph0.num_nodes() == 0);
   Eigen::Map<Eigen::Array<std::int32_t, Eigen::Dynamic, Eigen::Dynamic,
                           Eigen::RowMajor>>
       _dofmap(dofmap.data(), node_graph0.num_nodes(),
               dofmap.rows() / node_graph0.num_nodes());
-  return {std::move(index_map), element_block_size,
+  return {std::move(index_map), element_dof_layout.block_size(),
```
[the remark this answers is not in the document]
Comment (garth-wells) [reply cpp/dolfinx/fem/DofMapBuilder.cpp:559]: There's no guarantee that the dofmap builder uses the block size. One might want to 'unroll'.

File: cpp/dolfinx/la/SparsityPattern.cpp (hunk 4 above)
Comment (garth-wells) [reply cpp/dolfinx/la/SparsityPattern.cpp:38]: The function concatenates sparsity patterns - the input sparsity patterns can have bs > 1, but the  concatenated pattern always has bs = 1. A future optimisation could handle the case where all input sparsity patterns have the same bloc sizes.

**Review by chrisrichardson — APPROVED** [review]

**Review by michalhabera — APPROVED** [review]

**Pull request merged**